File: leetcode/15.3-sum.go
package main

import "sort"

/*
 * @lc app=leetcode id=15 lang=golang
 *
 * [15] 3Sum
 */

// @lc code=start
func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	ans := make([][]int, 0)
	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		all_two_sums := twoSum(nums[i+1:], -nums[i])
		for _, two_sums_idx := range all_two_sums {
			ans = append(ans, []int{nums[i], two_sums_idx[0], two_sums_idx[1]})
		}
	}
	return ans
}

func twoSum(nums []int, target int) [][]int {
	ans := make([][]int, 0)
	j := len(nums) - 1
	for i := 0; i < len(nums); i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		for i < j && nums[i]+nums[j] > target {
			j--
		}
		if i < j && nums[i]+nums[j] == target {
			ans = append(ans, []int{nums[i], nums[j]})
		}
	}
	return ans
}

// @lc code=end

File: leetcode/15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if(nums.size() < 3) return {};
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < nums.size() - 2; i++) {
            if (i >= 1 && nums[i] == nums[i - 1]) continue;
            for (auto vec : twoSums(nums, i + 1, nums.size() - 1, -nums[i]))
                res.push_back({nums[i], vec[0], vec[1]});
        }
        return res;
    }
private:
    vector<vector<int>> twoSums(vector<int>& nums, int l, int r, int target) {
        vector<vector<int>> res;
        for (int i = l, j = r; i < nums.size() - 1; i++) {
            if (i - 1 >= l && nums[i] == nums[i - 1]) continue;
            while (i < j && nums[i] + nums[j] > target) j--;
            if (i < j && nums[i] + nums[j] == target) 
                res.push_back({nums[i], nums[j]});
        }
        return res;
    }
};
// @lc code=end

